zookeeper-server/src/main/java/org/apache/zookeeper/server/embedded/EmbeddedZooKeeper.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.server.embedded;
+
+import org.apache.zookeeper.util.ServiceUtils;
+
+/**
+ * Factory methods for starting ZooKeeper servers.
+ */
+public final class EmbeddedZooKeeper {
+    private EmbeddedZooKeeper() {
+        // Prevent this class from being instantiated.
+    }
+
+    /**
+     * Starts a ZooKeeper server in production mode, meaning the JVM will exit when the ZooKeeper
+     * server exits.
+     * @param config The {@link EmbeddedConfig}
+     * @return A handle to the ZooKeeper server
+     */
+    static ZooKeeperServerHandle startProd(final EmbeddedConfig config) {
+        ServiceUtils.setSystemExitProcedure(ServiceUtils.SYSTEM_EXIT);
+        return new QuorumPeerMainHandle(config);
+    }
+
+    /**
+     * Starts a ZooKeeper server in testing mode, meaning the JVM will not exit when the ZooKeeper
+     * server exits.
+     * @param config The {@link EmbeddedConfig}
+     * @return A handle to the ZooKeeper server
+     */
+    static ZooKeeperServerHandle startTest(final EmbeddedConfig config) {
```

Comment:
having two methods here is problematic.
why do not you add a property to EmbeddedConfig ?

because now you have only one difference (the SYSTEM_EXIT/LOG_ONLY stuff)  but when you will have more differences ?

Reply:
@eolivelli (this is my work account, but it's still StuporHero talking)

The reason I chose this design is so that the code clearly communicates the intention of the writer, both in terms of how the code reads and in terms of the javadoc. If I moved it to a property somewhere, the method call that starts the server would no longer communicate to readers of the code how they should expect the server to exit which is information that is most relevant when the server is started. Instead of getting an informative javadoc, they would get a generic statement like "starts a server" that at best forces someone to hunt down the decision somewhere less relevant and at worst totally obscures that there's a potential difference. I'm actually not convinced these are the best names for the methods, but that at least should give some insight into my rationale.

I also didn't add a property to `EmbeddedConfig` because of the separation of concerns. `EmbeddedConfig` is concerned with producing well-formed configuration files. Since deciding how the system should exit is not set in configuration files, including it in `EmbeddedConfig` would violate the separation of concerns.

As for unanticipated future change, the API can be redesigned in the future with backwards compatibility exactly as I'm doing with your design, so I'd rather design the API to work with zookeeper as it is today rather than how I think it might be tomorrow. It's hard to predict the future, so it is rarely helpful to try to write code in anticipation of it.

I hope my rationale makes sense. What do you think?

Reply:
@mjulianotq
Can you please start a discussion on dev@zookeeper.apache.org ?

this would be an API change and usually we start a discussion in order to let the community participate (not everyone is following PR comments)